metagpt/actions/ml_action.py

```diff
@@ -0,0 +1,102 @@
+import json
+from typing import List, Tuple
+
+from metagpt.actions import Action
+from metagpt.actions.write_analysis_code import WriteCodeWithTools
+from metagpt.prompts.ml_action import (
+    GENERATE_CODE_PROMPT,
+    ML_TOOL_USAGE_PROMPT,
+    PRINT_DATA_COLUMNS,
+    UPDATE_DATA_COLUMNS,
+)
+from metagpt.prompts.write_analysis_code import CODE_GENERATOR_WITH_TOOLS
+from metagpt.schema import Message, Plan
+from metagpt.utils.common import CodeParser, create_func_call_config, remove_comments
+
+
+class WriteCodeWithToolsML(WriteCodeWithTools):
+    async def run(
+        self,
+        context: List[Message],
+        plan: Plan = None,
+        column_info: str = "",
+        **kwargs,
+    ) -> Tuple[List[Message], str]:
+        # prepare tool schemas and tool-type-specific instruction
+        tool_schemas, tool_type_usage_prompt = await self._prepare_tools(plan=plan)
+
+        # ML-specific variables to be used in prompt
+        code_steps = plan.current_task.code_steps
+        finished_tasks = plan.get_finished_tasks()
+        code_context = [remove_comments(task.code) for task in finished_tasks]
+        code_context = "\n\n".join(code_context)
+
+        # prepare prompt depending on tool availability & LLM call
+        if tool_schemas:
+            prompt = ML_TOOL_USAGE_PROMPT.format(
```

Comment:
Why do you need to print the tool here? Logically it should either be used as a system_prompt or in a specified field, not in the prompt?

Reply:
- There are about twenty tools, if we all put them along with their schema in system prompt, it will very long. Therefore, we select a few of them dynamically in each task, this makes llm easier to choose and consumes fewer tokens. 
- We could not use the tool_choice field of openai api, because it that case, we only get filled arguments instead of a code block using the tools